### PR TITLE
GRPC Integration in GCSFuse

### PR DESCRIPTION
**Pull request by raj-prince**
### Description
Adding support in gcsfuse to communicate with GCS via gRPC.

- Added new value for `client-protocol` flags to mount gcsfuse with gRPC protocol. you need to pass --client-protocol=grpc
- Added new flag `grcp-conn-pool-size` to control the number of channel used in gRPC protocol client.
- Added a new library "googledirectpath" which works as google-c2p resolver.
- Refactored current NewStorageClient method to create both http & grpc client based on the passed value for `--client-protocol`
- Added unit test for the newly added method in storage package.

[Testing in progress]
1. Unit test failed with google: default credential not found.
2. Facing issue in update object method: generation not supported.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - Verified manually normal operations.
3. Unit tests - automation.
4. Integration tests -

### CODE REVIEW

File: internal/storage/storageutil/custom_retry.go [hunk 1]
```diff
@@ -25,6 +27,10 @@ func ShouldRetry(err error) (b bool) {
 		return
 	}
 
+	if st, ok := status.FromError(err); ok && st.Code() == codes.ResourceExhausted {
```
Comment (vadlakondaswetha) [inline internal/storage/storageutil/custom_retry.go:30]: is this a temporary change?

File: internal/storage/storage_handle.go [hunk 2]
```diff
@@ -51,12 +55,22 @@ type StorageClientConfig struct {
...
 	RetryMultiplier     float64
 	UserAgent           string
+	GRPCConnPoolSize    int
 }
 
-// NewStorageHandle returns the handle of Go storage client containing
-// customized http client. We can configure the http client using the
-// storageClientConfig parameter.
-func NewStorageHandle(ctx context.Context, clientConfig StorageClientConfig) (sh StorageHandle, err error) {
+func createGRPCClientHandle(ctx context.Context, clientConfig StorageClientConfig) (sc *storage.Client, err error) {
```
Comment (vadlakondaswetha) [inline internal/storage/storage_handle.go:61]: unit tests?

File: flags.go [hunk 3]
```diff
@@ -267,6 +267,13 @@ func newApp() (app *cli.App) {
 					" in the meantime using the same mount, since we are not refreshing the cache, it will still return nil.",
 			},
 
+			cli.IntFlag{
+				Name:  "grpc-conn-pool-size",
+				Value: 4,
+				Usage: "Number of gRPC channel. Normally, it's same as number of tcp connection, but in case of" +
+					"direct-path traffic it's hard to determine number of tcp-connections in one grpc channel.",
```
Comment (vadlakondaswetha) [inline flags.go:274]: the description is not clear. Are you saying one tcp connection is same as one gRPC channel?

**Review by vadlakondaswetha — CHANGES_REQUESTED** [review]

File: flags.go (hunk 3 above)
Comment (raj-prince) [reply flags.go:274]: Yes, for cloud-path number of gRPC channel is same as number of tcp connections.

File: internal/storage/storageutil/custom_retry.go (hunk 1 above)
Comment (raj-prince) [reply internal/storage/storageutil/custom_retry.go:30]: Yes, adding a todo to remove the change later.

File: internal/storage/storage_handle.go (hunk 2 above)
Comment (raj-prince) [reply internal/storage/storage_handle.go:61]: Added.

File: flags.go (hunk 3 above)
Comment (vadlakondaswetha) [reply flags.go]: what is a good value for direct-path?

**Comment by raj-prince** [comment]
Not required, closing this.

**Pull request closed without merging**